ip_list.go

```diff
@@ -0,0 +1,445 @@
+func (api *API) DeleteIPListItemsAsync(ctx context.Context, id string, items IPListItemDeleteRequest) (
+	IPListItemDeleteResponse, error) {
+	uri := fmt.Sprintf("/accounts/%s/rules/lists/%s/items", api.AccountID, id)
+	res, err := api.makeRequestContext(ctx, http.MethodDelete, uri, items)
+	if err != nil {
+		return IPListItemDeleteResponse{}, errors.Wrap(err, errMakeRequestError)
+	}
+
+	result := IPListItemDeleteResponse{}
+	if err := json.Unmarshal(res, &result); err != nil {
+		return IPListItemDeleteResponse{}, errors.Wrap(err, errUnmarshalError)
+	}
+
+	return result, nil
+}
+
+// DeleteIPListItems removes specific Items of an IP List by their ID synchronously and returns the current set
+// of IP List Items
+func (api *API) DeleteIPListItems(ctx context.Context, id string, items IPListItemDeleteRequest) (
+	[]IPListItem, error) {
+	result, err := api.DeleteIPListItemsAsync(ctx, id, items)
+	if err != nil {
+		return []IPListItem{}, err
+	}
+
+	err = api.pollIPListBulkOperation(ctx, result.Result.OperationID)
+	if err != nil {
+		return []IPListItem{}, err
+	}
+
+	return api.ListIPListItems(ctx, id)
+}
+
+// GetIPListItem returns a single IP List Item
+// API reference: https://api.cloudflare.com/#rules-lists-get-list-item
+func (api *API) GetIPListItem(ctx context.Context, listID string, id string) (IPListItem, error) {
+	uri := fmt.Sprintf("/accounts/%s/rules/lists/%s/items/%s", api.AccountID, listID, id)
+	res, err := api.makeRequestContext(ctx, http.MethodGet, uri, nil)
+	if err != nil {
+		return IPListItem{}, errors.Wrap(err, errMakeRequestError)
+	}
+
+	result := IPListItemsGetResponse{}
+	if err := json.Unmarshal(res, &result); err != nil {
+		return IPListItem{}, errors.Wrap(err, errUnmarshalError)
+	}
+
+	return result.Result, nil
+}
+
+// GetIPListBulkOperation returns the status of a bulk operation
+// API reference: https://api.cloudflare.com/#rules-lists-get-bulk-operation
+func (api *API) GetIPListBulkOperation(ctx context.Context, id string) (IPListBulkOperation, error) {
+	uri := fmt.Sprintf("/accounts/%s/rules/lists/bulk_operations/%s", api.AccountID, id)
+	res, err := api.makeRequestContext(ctx, http.MethodGet, uri, nil)
+	if err != nil {
+		return IPListBulkOperation{}, errors.Wrap(err, errMakeRequestError)
+	}
+
+	result := IPListBulkOperationResponse{}
+	if err := json.Unmarshal(res, &result); err != nil {
+		return IPListBulkOperation{}, errors.Wrap(err, errUnmarshalError)
+	}
+
+	return result.Result, nil
+}
+
+// this is a helper method to implement synchronous behavior for some asynchronous endpoints.
+// bulk-operation status can be either pending, running, failed or completed
+func (api *API) pollIPListBulkOperation(ctx context.Context, id string) error {
+	var i uint8
+	for i = 0; i < 16; i++ {
+		time.Sleep(0x1 << uint8(math.Ceil(float64(i/2))) * time.Second)
+
+		bulkResult, err := api.GetIPListBulkOperation(ctx, id)
+		if err != nil {
+			return err
+		}
+
+		switch bulkResult.Status {
```

Comment:
I'm not familiar enough with the internals of this functionality so feel free to educate but do we want a `default` case on this switch to account for any unknown exceptions or statuses that are returned? I'm not too tied to it, but it is a convention that has saved me a number of times in the past for unexpected scenarios.

Reply:
The bulk operation should return pending, running, completed or failed. But it's a very good point handling a default case here if something unexpected happens. I've added this.